### ISSUE TITLE
StringIndexOutOfBoundsException at LiftoverVcf

### ISSUE THREAD

**Issue by C37H41N2O6** [broadinstitute/picard#1933]
I'm attempting to liftover from GRCh38 to JG2.

GATK command is:

gatk LiftoverVcf \
    -I ${PWD}/autosome_GRCh38.vcf \
    -O ${PWD}/autosome.vcf \
    --CHAIN ${PWD}/GRCh38_to_JG2.chain \
    --REJECT ${PWD}/rejected_autosome.vcf \
    -R ${PWD}/jg.fa

autosome_GRCh38.vcf was downloaded from
https://jmorp.megabank.tohoku.ac.jp/downloads/tommo-54kjpn-20230626-af_snvindelall
(direct link:
https://jmorp.megabank.tohoku.ac.jp/datasets/tommo-54kjpn-20230626-af_snvindelall/files/tommo-54kjpn-20230626r3-GRCh38-af-autosome.vcf.gz
)

GRCh38_to_JG2.chain & jg.fa were downloaded from
https://jmorp.megabank.tohoku.ac.jp/downloads/tommo-jg2.1.0-20211208

(direct link:
https://jmorp.megabank.tohoku.ac.jp/datasets/tommo-jg2.1.0-20211208/files/GRCh38.p13.genome_to_JG2.1.0.chain
https://jmorp.megabank.tohoku.ac.jp/datasets/tommo-jg2.1.0-20211208/files/jg2.1.0.fa.gz
)

I get many warnings and it terminates with a String index out of range error. Any help is appreciated.

The tail end of the output follows:

INFO    2023-12-21 11:16:11     LiftOver        Interval chr17:22477226-22477227 failed to match chain 2622 because intersection length 1 < minMatchSize 2.0 (0.5 < 1.0)
INFO    2023-12-21 11:16:11     LiftOver        Interval chr17:22477820-22477823 failed to match chain 2622 because intersection length 2 < minMatchSize 4.0 (0.5 < 1.0)
INFO    2023-12-21 11:16:11     LiftOver        Interval chr17:22477821-22477824 failed to match chain 2622 because intersection length 1 < minMatchSize 4.0 (0.25 < 1.0)
INFO    2023-12-21 11:16:11     LiftOver        Interval chr17:22509801-22509802 failed to match chain 2605 because intersection length 1 < minMatchSize 2.0 (0.5 < 1.0)
[Thu Dec 21 11:16:12 GMT 2023] picard.vcf.LiftoverVcf done. Elapsed time: 169.33 minutes.
Runtime.totalMemory()=31088181248
To get help, see http://broadinstitute.github.io/picard/index.html#GettingHelp
java.lang.StringIndexOutOfBoundsException: offset -1, count 3, length 5811
        at java.base/java.lang.String.checkBoundsOffCount(String.java:4587)
        at java.base/java.lang.String.<init>(String.java:397)
        at htsjdk.samtools.util.StringUtil.bytesToString(StringUtil.java:301)
        at picard.vcf.LiftoverVcf.tryToAddVariant(LiftoverVcf.java:558)
        at picard.vcf.LiftoverVcf.doWork(LiftoverVcf.java:453)
        at picard.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:289)
        at org.broadinstitute.hellbender.cmdline.PicardCommandLineProgramExecutor.instanceMain(PicardCommandLineProgramExecutor.java:37)
        at org.broadinstitute.hellbender.Main.runCommandLineProgram(Main.java:160)
        at org.broadinstitute.hellbender.Main.mainEntry(Main.java:203)
        at org.broadinstitute.hellbender.Main.main(Main.java:289)
Using GATK jar /gatk/gatk-package-4.4.0.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /gatk/gatk-package-4.4.0.0-local.jar LiftoverVcf -I /data/group1/z40265r/jg2/autosome_GRCh38.vcf -O /data/group1/z40265r/jg2/autosome.vcf --CHAIN /data/group1/z40265r/jg2/GRCh38_to_JG2.chain --REJECT /data/group1/z40265r/jg2/rejected_autosome.vcf -R /data/group1/z40265r/jg2/jg.fa

**Comment by rickymagner** [comment]
Hi, I was able to track down exactly which line was causing this issue, and I think I know what the problem is. There's an edge case in how we do liftover that was not considered, and is hit by one line is the vcf. It occurs here:
```
chr17   22533036        rs2011471686    GTA     G       6399.12 PASS ... <omitted>
```

The reason is that the chain file positions this a few bases away from the start of where it maps. Here is the relevant line in the chain file:
```
chain   4900    chr17   83257441        +       22527230        22533041        unplaced_85     5811    -       0       5811    2535
```

There a tandem repeat of two bases that gets contracted in this variant, and when it maps to `unplaced_85` it gets left aligned in our algorithm to the very beginning of the contig in an edge case that we don't handle properly. Thus we try to create a variant with `POS = 0`, and based on how certain strings get constructed in the code, a byte array tries to get accessed with index `POS - 1 = -1` which leads to the cryptic exception you observed.

In short, the fastest "fix" for you would be to simply remove this line from your vcf if you find that tolerable. In the meantime, I'll make a separate issue with much more detail/evidence for our developers to try to work on fixing this in the longterm and providing a more helpful error message at least.

**Comment by C37H41N2O6** [comment]
Thank you for your excellent sleuthing work.
I won't be able to try it until mid-April, but appreciate it in advance.

**Comment by rickymagner** [comment]
Sounds good. For now I'll close this issue then as it's superseded by the bug report #1951.
